libicyque.c

```diff
@@ -963,9 +992,96 @@ icq_got_buddy_icon(IcyQueAccount *ia, JsonObject *obj, gpointer user_data)
 	g_dataset_destroy(buddy);
 }
 
+static void
+icq_get_chat_history(IcyQueAccount *ia, const gchar* chatId, const gchar* fromMsg, gint64 count, IcyQueProxyCallbackFunc callback, gpointer user_data)
+{
+	JsonObject* getHistoryParams = json_object_new();
+	json_object_set_string_member(getHistoryParams, "aimsid", ia->aimsid);
+	json_object_set_string_member(getHistoryParams, "lang", "en-US");
+	//FIXME: I have absolutely no idea where this comes from or what this is.
+	//TODO: the initial histDlgState event contains a patchVersion. Store this per conversation and use that..?
+	json_object_set_string_member(getHistoryParams, "patchVersion", "1");
+	json_object_set_string_member(getHistoryParams, "sn", chatId);
+	json_object_set_string_member(getHistoryParams, "fromMsgId", fromMsg);
+	json_object_set_int_member(getHistoryParams, "count", count);
+	
+	JsonObject *getHistoryRequest = icq_generate_robusto_request(ia, "getHistory", getHistoryParams);
+	const gchar* getHistoryRequestStr = json_object_to_string(getHistoryRequest);
+	json_object_unref(getHistoryRequest);
+	
+	icq_fetch_url_with_method(ia, "POST", ICQ_RAPI_SERVER, getHistoryRequestStr, callback, user_data);
+}
+
 static GList *valid_icyque_accounts = NULL;
 #define ICYQUE_ACCOUNT_IS_VALID(ia) (g_list_find(valid_icyque_accounts, (ia)) != NULL)
 
+static void
+icq_mark_message_as_read(IcyQueAccount *ia, const gchar *sn, const gchar *messageId)
+{
+	JsonObject *setDlgStateParams = json_object_new();
+	JsonArray *exclude = json_array_new();
+	json_object_set_string_member(setDlgStateParams, "aimSid", ia->aimsid);
+	json_object_set_array_member(setDlgStateParams, "exclude", exclude);
+	json_object_set_string_member(setDlgStateParams, "lastRead", messageId);
+	json_object_set_string_member(setDlgStateParams, "sn", sn);
+	
+	JsonObject *setDlgStateRequest = icq_generate_robusto_request(ia, "setDlgState", setDlgStateParams);
+	const gchar* setDlgStateRequestStr = json_object_to_string(setDlgStateRequest);
+	json_object_unref(setDlgStateRequest);
+	
+	icq_fetch_url_with_method(ia, "POST", ICQ_RAPI_SERVER, setDlgStateRequestStr, NULL, NULL);
+}
+
+static void
+icq_unread_message_load_cb(IcyQueAccount *ia, JsonObject *data, gpointer user_data)
+{
+	JsonObject *status = json_object_get_object_member(data, "status");
+	if(status && json_object_get_int_member(status, "code") == 20000) {
+	JsonObject *results = json_object_get_object_member(data, "results");
+		// Acquire persons first
+		JsonArray *persons = json_object_get_array_member(results, "persons");
+		const gchar* sn = NULL; // TODO: Support group chats here
+		gint i, len = json_array_get_length(persons);
+		if(len == 0) return;
+		if(len > 1) {
+			purple_connection_error(ia->pc, PURPLE_CONNECTION_ERROR_OTHER_ERROR, "Group-Conversations not yet supported. Please post the icyque input of the debug window in a github issue.");

```

Comment:
This will disconnect the whole account, rather than just displaying an error message.  Is that intended?

Reply:
My intention was, that this would raise the chance, that someone submits the necessary data to implement this. :grin:
Should I reduce it to a warning?

Reply:
I guess it's ok as long as there's no code path that could accidentally trigger it.  Would be a poor user experience if they got stuck in a login/logout loop